internal/url/url.go

```diff
@@ -0,0 +1,35 @@
+package url
+
+import (
+	"regexp"
+)
+
+var (
+	isSchemeRegExp   = regexp.MustCompile(`^[^:]+://`)
+	scpLikeUrlRegExp = regexp.MustCompile(`^(?:(?P<user>[^@]+)@)?(?P<host>[^:\s]+):(?:(?P<port>[0-9]{1,5})/)?(?P<path>[^\\].*)$`)
+)
+
+// MatchesScheme returns true if the given string matches a URL-like
+// format scheme.
+func MatchesScheme(url string) bool {
+	return isSchemeRegExp.MatchString(url)
+}
+
+// MatchesScpLike returns true if the given string matches an SCP-like
+// format scheme.
+func MatchesScpLike(url string) bool {
+	return scpLikeUrlRegExp.MatchString(url)
+}
+
+// FindScpLikeComponents returns the user, host, port and path of the
+// given SCP-like URL.
+func FindScpLikeComponents(url string) (user, host, port, path string) {
+	m := scpLikeUrlRegExp.FindStringSubmatch(url)
+	return m[1], m[2], m[3], m[4]
+}
+
+// IsLocalEndpoint returns true if the given URL string specifies a
+// local file endpoint.
+func IsLocalEndpoint(url string) bool {
+	return !MatchesScheme(url) && !MatchesScpLike(url)
```

Comment:
Could you provide an example of such "local endpoint", perhaps also in comment? Also, is it intended `file://` doesn't count as local?

Reply:
To be clear I didn't change anything here, it's supposed to do the same thing the endpoint parser already does when figuring out if it's a local endpoint or not.

It looks like git itself does support "file://", but it treats it like a network transfer: https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols

I doubt go-git handles "file://", but if it did, it would be in the `MatchesScheme` flow for endpoints, which I haven't studied.

Reply:
Sounds good; thanks!